### PR TITLE
[install] store resolved workspace path in lockfile

### LINKED ISSUE #3685: Fresh installation in a workspace fails with missing linked packages
### What version of Bun is running?

v0.6.15 (17903570)

### What platform is your computer?

Darwin 21.6.0 arm64 arm

### What steps can reproduce the bug?

Create a file `package.json` with:
```
{
  "name": "bun-workspaces",
  "module": "index.ts",
  "type": "module",
  "workspaces": [
    "packages/a"
  ],
  "dependencies": {
    "a": "workspace:*"
  }
}
```

Create a second file `packages/a/package.json` with:

```
{
  "name": "a",
  "version": "1.0.0"
}
```

Then run `bun install` in the root. This will correctly create `node_modules` with a linked package `a`, and a `bun.lockb` file.

Then simulate a fresh install: remove `node_modules` and run `bun install` again (`rm -fr **/node_modules; bun install`).

### What is the expected behavior?

When `bun.lockb` is committed to the repo and the project is a monorepo with linked workspaces, running `bun install` should just work.

### What do you see instead?

Instead, the second time you run it (i.e. with `bun.lockb` present but deleted `node_modules`) you get:

```
bun install v0.6.15 (17903570)


error: package "a" is not linked

To install a linked package:
   bun link my-pkg-name-from-package-json

Tip: the package name is from package.json, which can differ from the folder name.
```

### Additional information

image

### LINKED ISSUE #3958: Error installing packages from workspaces when using workspace protocol
### What version of Bun is running?

0.7.1

### What platform is your computer?

Darwin 21.6.0 arm64 arm

### What steps can reproduce the bug?

Create a `package.json` with

```
{
  "name": "bun-workspaces",
  "module": "index.ts",
  "type": "module",
  "workspaces": [
    "packages/a"
  ],
  "dependencies": {
    "a": "workspace:*"
  }
}
```

And a `packages/a/package.json` with:

```
{
  "name": "a",
  "version": "1.0.0"
}
```

Then run `bun install`.

### What is the expected behavior?

It installs without errors. `node_modules/a` is linked to `packages/a`.

### What do you see instead?

Got an error:

```
bun install v0.7.1 (53cc4df1)


error: workspace dependency "a" not found

Searched in "./*"

Workspace documentation: https://bun.sh/docs/install/workspaces
```

### Additional information

Looks related to the `"workspace:*"` dependency, because depending on just `"*"` seems to work.

### PR DESCRIPTION
**Pull request by alexlamsl**
- resolve cwd consistently
- speed up package diff
- update `test/bun.lockb`

fixes #3685
fixes #3958

### What does this PR do?

- [ ] Documentation or TypeScript types (it's okay to leave the rest blank in this case)
- [x] Code changes

### How did you verify your code works?

I wrote automated tests

If Zig files changed:

- [x] I checked the lifetime of memory allocated to verify it's (1) freed and (2) only freed when it should be
- [x] I or my editor ran `zig fmt` on the changed files
- [x] I included a test for the new code, or an existing test covers it
- [ ] JSValue used outside outside of the stack is either wrapped in a JSC.Strong or is JSValueProtect'ed

### CODE REVIEW

File: src/install/lockfile.zig [hunk 1]
```diff
@@ -2651,20 +2648,22 @@ pub const Package = extern struct {
                             ),
                         ),
                     );
-                    defer dependency_version.value.workspace = path;
+                    defer {
+                        dependency_version.literal = path;
+                        dependency_version.value.workspace = path;
```
Comment (Jarred-Sumner) [inline src/install/lockfile.zig]: This is not the absolute path right

This is relative to the root package.json?

Could we have an assertion here that asserts its not an absolute path

**Review by Jarred-Sumner — APPROVED** [review]
LGTM after adding assertion

File: src/install/lockfile.zig (hunk 1 above)
Comment (alexlamsl) [reply src/install/lockfile.zig]: > This is relative to the root package.json?

Yes, from the line directly above this:
https://github.com/oven-sh/bun/blob/76dee9312ba5201583ba9f8232b96397fa490808/src/install/lockfile.zig#L2640-L2653

> Could we have an assertion here that asserts its not an absolute path

Does checking `path[0] != '/'` cover this or is there a special function for that?

Comment (Jarred-Sumner) [reply src/install/lockfile.zig]: `std.debug.assert(!std.fs.path.isAbsolute(path)`

**Pull request merged**